File: crates/octobot_commons/tests/test_symbol_util.rs
use octobot_commons::enums::OptionTypes;
use octobot_commons::symbols::symbol::Symbol;
use octobot_commons::symbols::symbol_util::*;

#[test]
fn test_parse_symbol() {
    assert_eq!(parse_symbol("BTC/USDT"), Symbol::new("BTC/USDT"));
}

#[test]
fn test_merge_symbol() {
    assert_eq!(merge_symbol("BTC/USDT"), "BTCUSDT");
    assert_eq!(merge_symbol("BTC/USDT:USDT"), "BTCUSDT_USDT");
}

#[test]
fn test_merge_currencies_simple() {
    assert_eq!(
        merge_currencies("BTC", "USDT", None, None, None, None, None, None, None),
        "BTC/USDT"
    );
}

#[test]
fn test_merge_currencies_with_settlement() {
    assert_eq!(
        merge_currencies("BTC", "USDT", Some("BTC"), None, None, None, None, None, None),
        "BTC/USDT:BTC"
    );
}

#[test]
fn test_merge_currencies_custom_separators() {
    assert_eq!(
        merge_currencies(
            "BTC", "USDT", Some("XXX"), None, None, None, Some("g"), Some("d"), None
        ),
        "BTCgUSDTdXXX"
    );
}

#[test]
fn test_merge_currencies_option_put() {
    assert_eq!(
        merge_currencies(
            "will-bitcoin-replace-sha-256-before-2027",
            "USDC",
            Some("USDC"),
            Some("261231"),
            Some("0"),
            Some(OptionTypes::Put.value()),
            None,
            None,
            None,
        ),
        "will-bitcoin-replace-sha-256-before-2027/USDC:USDC-261231-0-P"
    );
}

#[test]
fn test_merge_currencies_option_custom_type() {
    assert_eq!(
        merge_currencies(
            "will-bitcoin-replace-sha-256-before-2027",
            "USDC",
            Some("USDC"),
            Some("261231"),
            Some("0"),
            Some("test"),
            None,
            None,
            None,
        ),
        "will-bitcoin-replace-sha-256-before-2027/USDC:USDC-261231-0-TEST"
    );
}

#[test]
fn test_merge_currencies_option_none_type() {
    assert_eq!(
        merge_currencies(
            "will-bitcoin-replace-sha-256-before-2027",
            "USDC",
            Some("USDC"),
            Some("261231"),
            Some("0"),
            None,
            None,
            None,
            None,
        ),
        "will-bitcoin-replace-sha-256-before-2027/USDC:USDC"
    );
}

#[test]
fn test_convert_symbol() {
    assert_eq!(
        convert_symbol("BTC-USDT", "-", None, false, false, false),
        "BTC/USDT"
    );
}

#[test]
fn test_is_symbol_default_separator() {
    assert!(is_symbol("BTC/USDT", None));
    assert!(is_symbol("ETH/USDT", None));
    assert!(is_symbol("BTC/USDT:USDT", None));
    assert!(!is_symbol("BTC", None));
    assert!(!is_symbol("USDT", None));
}

#[test]
fn test_is_symbol_dash_separator() {
    assert!(is_symbol("BTC-USDT", Some("-")));
    assert!(!is_symbol("BTC", Some("-")));
}

#[test]
fn test_is_symbol_colon_separator() {
    assert!(is_symbol("BTC/USDT:USDT", Some(":")));
    assert!(!is_symbol("BTC/USDT", Some(":")));
}

#[test]
fn test_is_symbol_pipe_separator() {
    assert!(is_symbol("BTC|USDT", Some("|")));
    assert!(!is_symbol("BTC", Some("|")));
}

#[test]
fn test_is_symbol_edge_cases() {
    assert!(!is_symbol("", Some("/")));
    assert!(is_symbol("/", Some("/")));
    assert!(is_symbol("BTC/USDT/ETH", Some("/")));
}

#[test]
fn test_is_usd_like_coin() {
    assert!(is_usd_like_coin("USDT"));
    assert!(is_usd_like_coin("USDC"));
    assert!(is_usd_like_coin("USD"));
    assert!(is_usd_like_coin("BUSD"));
    assert!(!is_usd_like_coin("BTC"));
    assert!(!is_usd_like_coin("ETH"));
    assert!(!is_usd_like_coin(""));
}

#[test]
fn test_get_most_common_usd_like_symbol() {
    let pairs = vec!["BTC/USDT", "ETH/USDT", "SOL/USDC"];
    let result = get_most_common_usd_like_symbol(&pairs);
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), "USDT");

    let pairs = vec!["BTC/USDC", "ETH/USDC", "SOL/USDT"];
    let result = get_most_common_usd_like_symbol(&pairs);
    assert!(result.is_ok());
    assert_eq!(result.unwrap(), "USDC");

    let empty: Vec<&str> = vec![];
    let result = get_most_common_usd_like_symbol(&empty);
    assert!(result.is_err());

    let no_usd = vec!["BTC/ETH", "SOL/BTC"];
    let result = get_most_common_usd_like_symbol(&no_usd);
    assert!(result.is_err());
}

File: packages/commons/tests/symbols/test_symbol_util.py
#  Drakkar-Software OctoBot-Commons
#  Copyright (c) Drakkar-Software, All rights reserved.
#
#  This library is free software; you can redistribute it and/or
#  modify it under the terms of the GNU Lesser General Public
#  License as published by the Free Software Foundation; either
#  version 3.0 of the License, or (at your option) any later version.
#
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
#  Lesser General Public License for more details.
#
#  You should have received a copy of the GNU Lesser General Public
#  License along with this library.
import octobot_commons.symbols


def test_parse_symbol():
    assert octobot_commons.symbols.parse_symbol("BTC/USDT") == octobot_commons.symbols.Symbol("BTC/USDT")


def test_merge_symbol():
    assert octobot_commons.symbols.merge_symbol("BTC/USDT") == "BTCUSDT"
    assert octobot_commons.symbols.merge_symbol("BTC/USDT:USDT") == "BTCUSDT_USDT"


def test_merge_currencies():
    assert octobot_commons.symbols.merge_currencies("BTC", "USDT") == "BTC/USDT"
    assert octobot_commons.symbols.merge_currencies("BTC", "USDT", "BTC") == "BTC/USDT:BTC"
    assert octobot_commons.symbols.merge_currencies("BTC", "USDT", settlement_asset="XXX", market_separator="g", settlement_separator="d") == "BTCgUSDTdXXX"
    assert (
        octobot_commons.symbols.merge_currencies(
            "will-bitcoin-replace-sha-256-before-2027",
            "USDC",
            "USDC",
            "261231",
            "0",
            octobot_commons.enums.OptionTypes.PUT.value
        )
        == "will-bitcoin-replace-sha-256-before-2027/USDC:USDC-261231-0-P"
    )
    assert (
        octobot_commons.symbols.merge_currencies(
            "will-bitcoin-replace-sha-256-before-2027",
            "USDC",
            "USDC",
            "261231",
            "0",
            "test"
        )
        == "will-bitcoin-replace-sha-256-before-2027/USDC:USDC-261231-0-TEST"
    )
    assert (
        octobot_commons.symbols.merge_currencies(
            "will-bitcoin-replace-sha-256-before-2027",
            "USDC",
            "USDC",
            "261231",
            "0",
            None
        )
        == "will-bitcoin-replace-sha-256-before-2027/USDC:USDC"
    )


def test_convert_symbol():
    assert octobot_commons.symbols.convert_symbol("BTC-USDT", symbol_separator="-") == "BTC/USDT"


def test_is_symbol():
    # Test with default separator (/)
    assert octobot_commons.symbols.is_symbol("BTC/USDT") is True
    assert octobot_commons.symbols.is_symbol("ETH/USDT") is True
    assert octobot_commons.symbols.is_symbol("BTC/USDT:USDT") is True
    assert octobot_commons.symbols.is_symbol("BTC") is False
    assert octobot_commons.symbols.is_symbol("USDT") is False
    assert octobot_commons.symbols.is_symbol("ETH") is False
    
    # Test with custom separator (-)
    assert octobot_commons.symbols.is_symbol("BTC-USDT", separator="-") is True
    assert octobot_commons.symbols.is_symbol("ETH-USDT", separator="-") is True
    assert octobot_commons.symbols.is_symbol("BTC", separator="-") is False
    assert octobot_commons.symbols.is_symbol("USDT", separator="-") is False
    
    # Test with custom separator (:)
    assert octobot_commons.symbols.is_symbol("BTC/USDT:USDT", separator=":") is True
    assert octobot_commons.symbols.is_symbol("BTC/USDT", separator=":") is False
    assert octobot_commons.symbols.is_symbol("BTC", separator=":") is False
    
    # Test with custom separator (|)
    assert octobot_commons.symbols.is_symbol("BTC|USDT", separator="|") is True
    assert octobot_commons.symbols.is_symbol("BTC", separator="|") is False
    
    # Test edge cases
    assert octobot_commons.symbols.is_symbol("", separator="/") is False
    assert octobot_commons.symbols.is_symbol("/", separator="/") is True
    assert octobot_commons.symbols.is_symbol("BTC/USDT/ETH", separator="/") is True


import pytest

import octobot_commons.constants as commons_constants


class TestTradingTypeFromSymbol:
    def test_spot_symbol_returns_spot(self):
        assert octobot_commons.symbols.trading_type_from_symbol("BTC/USDT") == commons_constants.CONFIG_EXCHANGE_SPOT

    def test_futures_symbol_returns_future(self):
        assert (
            octobot_commons.symbols.trading_type_from_symbol("BTC/USDT:USDT")
            == commons_constants.CONFIG_EXCHANGE_FUTURE
        )

    def test_option_symbol_returns_option(self):
        assert (
            octobot_commons.symbols.trading_type_from_symbol("BTC/USDT:USDT-211225-60000-P")
            == commons_constants.CONFIG_EXCHANGE_OPTION
        )

    def test_unknown_symbol_raises_value_error(self):
        with pytest.raises(ValueError):
            octobot_commons.symbols.trading_type_from_symbol("BTC")


class TestTradingTypeFromTradedSymbols:
    def test_all_spot_symbols_return_spot(self):
        assert octobot_commons.symbols.trading_type_from_traded_symbols(
            ["BTC/USDT", "ETH/USDT"]
        ) == commons_constants.CONFIG_EXCHANGE_SPOT

    def test_all_futures_symbols_return_future(self):
        assert octobot_commons.symbols.trading_type_from_traded_symbols(
            ["BTC/USDT:USDT", "ETH/USDT:USDT"]
        ) == commons_constants.CONFIG_EXCHANGE_FUTURE

    def test_empty_symbols_raises_value_error(self):
        with pytest.raises(ValueError):
            octobot_commons.symbols.trading_type_from_traded_symbols([])

    def test_mixed_symbol_types_raise_ambiguous_error(self):
        with pytest.raises(octobot_commons.errors.AmbiguousTradedSymbolsTradingTypeError):
            octobot_commons.symbols.trading_type_from_traded_symbols(["BTC/USDT", "BTC/USDT:USDT"])


class TestMergeSymbolNetworkDex:
    def test_strips_network_and_dex_suffix(self):
        assert octobot_commons.symbols.merge_symbol("SOL/USDT@solana!raydium") == "SOLUSDT"


class TestConvertSymbolBaseAndQuoteOnlyNetworkDex:
    def test_strips_network_and_dex_suffix(self):
        assert (
            octobot_commons.symbols.convert_symbol(
                "SOL/USDT@solana!raydium",
                symbol_separator="/",
                base_and_quote_only=True,
            )
            == "SOL/USDT"
        )
